pkg/populator/populator.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023 OnMetal authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package populator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/golang/groupcache/singleflight"
+	onmetalimage "github.com/onmetal/onmetal-image"
+	"github.com/onmetal/onmetal-image/oci/image"
+	"github.com/onmetal/onmetal-image/oci/remote"
+)
+
+type Provisioner interface {
+	CreateOsImage(ctx context.Context, imageId string, image image.Image) error
+	GetOsImage(ctx context.Context, imageId string) (bool, error)
+	DeleteOsImage(ctx context.Context, imageId string) error
+}
+
+func New(log logr.Logger, provisioner Provisioner) (*populator, error) {
+	var err error
+	p := &populator{
+		log:         log,
+		provisioner: provisioner,
+		populated:   map[string]string{},
+	}
+	p.registry, err = remote.DockerRegistry(nil)
```

Comment:
This should be passed in (e.g. via an `Options` struct) and not statically assigned.